### DISCUSSION TITLE
How are u rendering custom nodes into the graph view?

### DISCUSSION THREAD

**Discussion by BlitzVI** [jackvonhouse/custom-node-size#2] [Q&A]
Hey! Could you please explain to me how you are rendering custom nodes in the graph view? I'd like to know as i want to develop my own plugin that adds custom nodes to the graph view but i can't seem to figure out how. Any help is greatly appreciated!. You can contact me on my discord aswell as im much more active on discord. My discord id is bltz_v. Thank you!

**Comment by jackvonhouse** [comment]
@BlitzVI Hi! I don't render node directly. I just modify the `weight` param, which is then rendered with Obsidian using the `pixi.js` library.

If you want to add custom node, you must inspect `renderer` structure. To do this, press `Ctrl+Shift+i` to open the console. Then open the graph and execute this code in the console:
```js
this.app.workspace.getLeavesOfType("graph").first().view.renderer
```

I can also recommend [this](https://github.com/natefrisch01/Graph-Link-Types) repo, which renders label in lines between nodes.

**Comment by BlitzVI** [comment]
> @BlitzVI Hi! I don't render node directly. I just modify the `weight` param, which is then rendered with Obsidian using the `pixi.js` library.
> 
> If you want to add custom node, you must inspect `renderer` structure. To do this, press `Ctrl+Shift+i` to open the console. Then open the graph and execute this code in the console:
> 
> ```js
> this.app.workspace.getLeavesOfType("graph").first().view.renderer
> ```
> 
> I can also recommend [this](https://github.com/natefrisch01/Graph-Link-Types) repo, which renders label in lines between nodes.

ah alr i've already messed around with asking the exact same question in that repo but i still dont understand how i can add my own nodes just by looking at the src. If you could help me setup this renderer thing, i could try and mess around with it myself. Also is there any documentation that you are following? i have no clue what getleaf and such does. Also mb for making the issue ugly i dont use github much. Thank you for the reply!

**Reply by jackvonhouse** [reply]
It would be better if you write exactly what your task is. What kind of nodes do you want to create?

**Reply by BlitzVI** [reply]
ah so what i want to do is add headings present in my notes as individual nodes in the graph view. When I link a heading in another note, it links that heading's node to the linked note and its respective heading. I've seen people request this feature in the forums and discord server for the past 2 years so i think i'd work well for my workflow and others' workflow aswell

**Reply by jackvonhouse** [reply]
I understand your case perfectly. But I think this problem is solved by changing the structure of your vault. You can store a note as a directory with different notes (headers):

```
.
└── My note/
    ├── Content/
    │   ├── Header 1
    │   ├── Header ...
    │   └── Header N
    └── My note.md
```

Then you refer to `Header 1`, `Header ...`, `Header N` in your `My note.md`:

```
# [[Header 1]]

# [[Header ...]]

# [[Header N]]
```

You can add `!` before each refer to embed `Header N` in `My note.md`:


```
![[Header 1]]

![[Header ...]]

![[Header N]]
```

That's exactly what I'm doing in my vaults. I think creating such a plugin seems unnecessary to me, as this can be solved by Obsidian tools.

**Reply by BlitzVI** [reply]
I've tried this approach aswell but the problem is for vaults that are really big you can't go around manually making heading notes as i think its a waste of time so i tried using a plugin that automatically separates notes and headings but it was too janky and didn't seem to  work for really long files perfectly. So i think the easiest way to combat this would be just to make headings appear as nodes in the graph view, that way the structure of our vault stays the same and you dont have the mess around making heading notes along with more graph  view functionality which people want aswell.

**Comment by jackvonhouse** [comment]
Okay, to create the plugin you need to do the following:

# Register event to open the graph
https://github.com/jackvonhouse/custom-node-size/blob/f69bcc59466b9980551315e9d57c2d4e7d986b00/src/main.ts#L6-L9

# Go through all nodes
https://github.com/jackvonhouse/custom-node-size/blob/f69bcc59466b9980551315e9d57c2d4e7d986b00/src/main.ts#L30-L34

## Read headers for each node

```js
const headers = fileCache?.headings;
```

`headings` has this struct:
```json
[
    {
        "heading": "Header",
        "level": 1
    }
]
```

## Add new nodes

There is no logic here related to header inheritance. All headers will inherit from the parent note. For other behavior you can convert `headers` into a tree or develop your own algorithm (ChatGPT to help).

```js
let newNodes = [];

nodes.forEach(node => {
    // ...

    headers.forEach(header => {
        // Write your deep copy algorithm for clone the object.
        let newNode = clone(node);

        // This is needed to jump to a given header when clicking on a node.
        // Example: "Programming/JavaScript.md#Types"
        const headerPrefix = '#'.repeat(header.level);
        const headerName = header.heading;
        newNode.id = `${node.id}${headerPrefix}${headerName}`;
        newNode.reverse = {
            // Fields associated with the root note.
            // Explore them for yourself.
        };

        newNodes.push(newNode);
    });
});

nodes.push(newNodes);
```

**Reply by jackvonhouse** [reply]
You can also avoid adding a reference to the header in `newNode.id`. Instead, you can override `onClick` in `newNode` to something like:
```js
this.app.workspace.openLinkText(`${node.id}${headerPrefix}${headerName}`, '')
```

**Reply by BlitzVI** [reply]
oh wow tysm for the help! I have exams from tmrw so i might not be able to implement this rn but i'll be sure to use this knowledge when ever i get to actually making the plugin. Again tysm for the help!
